0-module-and-module-source.emu

```diff
@@ -0,0 +1,661 @@
+                <dd>It creates a Source Text Module Record based upon the result of a previously parsed _sourceText_ as a |Module| associated to the provided Module Source Record.</dd>
+            </dl>
+            <emu-alg>
+              1. Assert: _moduleSource_ is a Module Source Record.
+              1. Let _async_ be _sourceModule_.[[HasTLA]].
+              1. Let _body_ be _sourceModule_.[[ECMAScriptCode]].
+              1. Let _requestedModules_ be _sourceModule_.[[RequestedModules]].
+              1. Let _importEntries_ be _sourceModule_.[[ImportEntries]].
+              1. Let _localExportEntries_ be _sourceModule_.[[LocalExportEntries]].
+              1. Let _indirectExportEntries_ be _sourceModule_.[[IndirectExportEntries]].
+              1. Let _starExportEntries_ be _sourceModule_.[[StarExportEntries]].
+              1. Let _hostDefined_ be _sourceModule_.[[HostDefined]].
+              1. Let _realm_ be the current Realm Record.
+              1. Return Source Text Module Record { [[Realm]]: _realm_, [[Environment]]: ~empty~, [[Namespace]]: ~empty~, [[CycleRoot]]: ~empty~, [[HasTLA]]: _async_, [[AsyncEvaluation]]: *false*, [[TopLevelCapability]]: ~empty~, [[AsyncParentModules]]: &laquo; &raquo;, [[PendingAsyncDependencies]]: ~empty~, [[Status]]: ~unlinked~, [[EvaluationError]]: ~empty~, [[HostDefined]]: _hostDefined_, [[ECMAScriptCode]]: _body_, [[Context]]: ~empty~, [[ImportMeta]]: ~empty~, [[RequestedModules]]: _requestedModules_, [[ImportEntries]]: _importEntries_, [[LocalExportEntries]]: _localExportEntries_, [[IndirectExportEntries]]: _indirectExportEntries_, [[StarExportEntries]]: _starExportEntries_, [[DFSIndex]]: ~empty~, [[DFSAncestorIndex]]: ~empty~, [[ModuleInstance]]: *undefined* }.
+            </emu-alg>
+        </emu-clause>
+
+        <emu-clause id="sec-resolvemodulerecorddependency" type="abstract operation">
+            <h1>
+                ResolveModuleRecordDependency (
+                    _moduleRecord_: a Module Record,
+                    _specifier_: a |ModuleSpecifier| String,
+                    _promiseCapability_: a PromiseCapability Record,
+                ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>It provides the concrete Module Record subclass instance that corresponds to _specifier_ occurring within the module represented by _moduleRecord_.</dd>
+            </dl>
+            <emu-alg>
+              1. Assert: _moduleRecord_ is a Module Record.
+              1. Assert: _moduleRecord_.[[ModuleInstance]] is an Object.
+              1. Assert: ResolveModuleRecordDependency has not been invoked with _moduleRecord_ and _specifier_ pair.
+              1. Let _moduleInstance_ be _moduleRecord_.[[ModuleInstance]].
+              1. Let _importHook_ be _moduleInstance_.[[ImportHook]].
+              1. Let _completion_ be Completion(Call(_importHook_, *undefined*, « _specifier_ »).
```

Comment:
If we want to ensure sync resolution possibilities, perhaps we do need to support `Promise<Instance> | Instance` as the API rather here?

Reply:
that is tricky as always... because the instance might have a `.then` on them, and then what?

Reply:
Pass the result to `PromiseResolve(_here_)` and it gets unified? 🤔

Reply:
Can we define the `Instance` JS object to explicitly never have a `then` property?

Reply:
> Can we define the Instance JS object to explicitly never have a then property?

@guybedford  No, I don't think we should do that. In fact `Module` is subclass-able.

> Pass the result to PromiseResolve(_here_) and it gets unified?

@Jack-Works that's precisely what I do in the following steps.